contracts/0.8.9/lib/WithdrawalRequests.sol

```diff
@@ -0,0 +1,122 @@
+     *      This allows withdrawal of any balance exceeding 32 ETH (e.g., if a validator has 35 ETH, up to 3 ETH can be withdrawn).
+     *      However, the protocol enforces a minimum balance of 32 ETH per validator, even if a higher amount is requested.
+     * @param pubkeys An array of public keys for the validators requesting withdrawals.
+     * @param amounts An array of corresponding withdrawal amounts for each public key.
+     */
+    function addPartialWithdrawalRequests(
+        bytes[] calldata pubkeys,
+        uint64[] calldata amounts
+    ) internal {
+        uint256 keysCount = pubkeys.length;
+        if (keysCount != amounts.length) {
+            revert MismatchedArrayLengths(keysCount, amounts.length);
+        }
+
+        uint64[] memory _amounts = new uint64[](keysCount);
+        for (uint256 i = 0; i < keysCount; i++) {
+            if (amounts[i] == 0) {
+                revert PartialWithdrawalRequired(pubkeys[i]);
+            }
+
+            _amounts[i] = amounts[i];
+        }
+
+        _addWithdrawalRequests(pubkeys, _amounts);
+    }
+
+    /**
+     * @dev Retrieves the current withdrawal request fee.
+     * @return The minimum fee required per withdrawal request.
+     */
+    function getWithdrawalRequestFee() internal view returns (uint256) {
+        (bool success, bytes memory feeData) = WITHDRAWAL_REQUEST.staticcall("");
+
+        if (!success) {
+            revert WithdrawalRequestFeeReadFailed();
+        }
+
+        return abi.decode(feeData, (uint256));
+    }
+
+    function _addWithdrawalRequests(
+        bytes[] calldata pubkeys,
+        uint64[] memory amounts
+    ) internal {
+        uint256 keysCount = pubkeys.length;
+        if (keysCount == 0) {
+            revert NoWithdrawalRequests();
+        }
+
+        uint256 minFeePerRequest = getWithdrawalRequestFee();
+        if (minFeePerRequest * keysCount > msg.value) {
+            revert FeeNotEnough(minFeePerRequest, keysCount, msg.value);
+        }
+
+        uint256 feePerRequest = msg.value / keysCount;
+        uint256 unallocatedFee = msg.value % keysCount;
+        uint256 prevBalance = address(this).balance - msg.value;
+
+
+        for (uint256 i = 0; i < keysCount; ++i) {
+            bytes memory pubkey = pubkeys[i];
+            uint64 amount = amounts[i];
+
+            if(pubkey.length != 48) {
+                revert InvalidPubkeyLength(pubkey);
+            }
+
+            uint256 feeToSend = feePerRequest;
+
+            if (i == keysCount - 1) {
+                feeToSend += unallocatedFee;
+            }
+
+            bytes memory callData = abi.encodePacked(pubkey, amount);
+            (bool success, ) = WITHDRAWAL_REQUEST.call{value: feeToSend}(callData);
+
+            if (!success) {
+                revert WithdrawalRequestAdditionFailed(pubkey, amount);
+            }
+
+            emit WithdrawalRequestAdded(pubkey, amount);
+        }
+
+        assert(address(this).balance == prevBalance);
```

Comment:
Is it possible to calculate the fee exactly at the moment of signing the tx? Shouldn't we ease the constraint and add some refund capability here?

Reply:
Great question. 

Calls to the system contract require a fee payment defined by the current contract state. Overpaid fees are not returned to the caller. It is not generally possible to compute the exact required fee amount ahead of time. 

When adding a withdrawal request from a contract, the contract can perform a read operation to check for the current fee and then pay an exact amount.

When adding a withdrawal request from EOA request may result in overpayment of fees. But the gas cost of returning the overage would likely be higher than the overage itself. 

Proposed in this PR implementation does not imply any restriction on the caller, but makes sure that the withdrawal credential balance (accounting) will not be affected.

In the proposed [Triggerable Withdrawal](https://hackmd.io/@lido/SJKVeT990) implementation, the dedicated bot will provide the fee within the withdrawal request transaction and, a small overpayment is possible. This is the simplest approach, which is currently considered for [Triggerable Withdrawal](https://hackmd.io/@lido/SJKVeT990)  implementation.

Reply:
Good point, I agree. The fee allocation strategy should be decoupled from the withdrawals library, allowing each contract to determine its own approach based on its design and requirements. To address these changes, I suggest adding a new `totalWithdrawalFee` parameter, which represents the total amount that will be spent on the EIP-7002 withdrawal requests.

The following pseudo-code demonstrates the approach:
```solidity
library WithdrawalRequests {
    function addWithdrawalRequests(
        bytes[] calldata pubkeys,
        uint64[] memory amounts,
        uint256 totalWithdrawalFee
    ) internal {
        // Implementation that uses totalWithdrawalFee to cover EIP-7002 withdrawal requests
        ...
    }

    function getWithdrawalRequestFee() internal pure returns (uint256) {
        // Returns the minimum required fee per request
        ...
    }
}

contract A {
    function addWithdrawalRequests(
        bytes[] calldata pubkeys,
        uint64[] calldata amounts
    ) external payable {
        // Use the entire sent amount (msg.value) as the total fee for withdrawal requests
        uint256 totalWithdrawalFee = msg.value;
        WithdrawalRequests.addWithdrawalRequests(pubkeys, amounts, totalWithdrawalFee);
    }
}

contract B {
    function addWithdrawalRequests(
        bytes[] calldata pubkeys,
        uint64[] calldata amounts
    ) external {
        // Use the minimum required fee per request
        uint256 minFeePerRequest = WithdrawalRequests.getWithdrawalRequestFee();
        uint256 totalWithdrawalFee = minFeePerRequest * pubkeys.length;
        WithdrawalRequests.addWithdrawalRequests(pubkeys, amounts, totalWithdrawalFee);
    }
}

// Additional contracts can implement other fee allocation strategies as needed.
```